File: ruby/run-length-encoding/run_length_encoding.rb
module BookKeeping
  VERSION = 3
end

class RunLengthEncoding
  # given a string, returns a run length encoded string
  def self.encode(str)
    count = 0
    char = ''
    out = ''

    str.chars.each do |alpha|
      count += 1
      next unless alpha != char
      out += append_encoded(char, count) unless char == ''
      char = alpha
      count = 0
    end
    out += append_encoded(char, count + 1)
  end

  # given a run length encoded string, returns the decoded string
  def self.decode(str)
    # use a string to store count to account for multi digit nums
    count = ''
    out = ''

    str.chars.each do |char|
      # build the count string
      if char =~ /\d/
        count += char
        next
      end
      # using the count and the current char, build the output string
      count = '1' if count == ''
      count.to_i.times do
        out += char
      end
      count = ''
    end
    out
  end

  def self.append_encoded(char, count)
    out = count > 1 ? count.to_s : ''
    out + char
  end
end

File: javascript/run-length-encoding/run-length-encoding.js
class RLE {
  // given a string, returns a run length encoded string
  static encode(str) {
    let count = 0;
    let char = '';
    let out = '';

    str.split('').forEach((alpha) => {
      count += 1;
      if (alpha === char) {
        return;
      }
      if (char !== '') {
        out += this.appendEncoded(char, count);
      }
      char = alpha;
      count = 0;
    });
    out += this.appendEncoded(char, count + 1);

    return out;
  }

  // given a run length encoded string, returns the decoded string
  static decode(str) {
    // use a string to store count to account for multi digit nums
    let count = '';
    let out = '';

    str.split('').forEach((char) => {
      // build the count string
      if (char.match(/\d/)) {
        count += char;
        return;
      }
      // handle single char runs
      count = count ? parseInt(count, 10) : 1;
      // using the count and the current char, build the output string
      for (let i = 0; i < count; i += 1) {
        out += char;
      }
      count = '';
    });

    return out;
  }

  // helper method to build the encoded string
  static appendEncoded(char, count) {
    return (count > 1 ? count.toString() : '') + char;
  }
}
module.exports = RLE;
